qemu/tests/usb_redir.py

```diff
@@ -0,0 +1,270 @@
+        rv_thread = utils_misc.InterruptedThread(os.system, (rv_args,))
+        rv_thread.start()
+        if not utils_misc.wait_for(_rv_connection_check, timeout, 60):
+            status = False
+            err_msg = "Fail to establish %s connection" % rv_binary
+        return (status, err_msg)
+
+    def boot_check(info):
+        """
+        boot info check
+        """
+        return re.search(info, vm.serial_console.get_stripped_output())
+
+    def _usb_dev_verify():
+        error_context.context("Check USB device in guest", logging.info)
+        if session.cmd_status(lsusb_cmd):
+            return False
+        return True
+
+    def _kill_rv_proc():
+        s, o = process.getstatusoutput("pidof %s" % rv_binary)
+        if not s:
+            process.getoutput("killall %s" % rv_binary)
+
+    def _get_usb_mount_point():
+        """ Get redirected USB stick mount point """
+        dmesg_cmd = "dmesg | grep 'Attached SCSI removable disk'"
+        s, o = session.cmd_status_output(dmesg_cmd)
+        if s:
+            test.error("Fail to get redirected USB stick in guest.")
+        dev = re.findall(r'\[(sd\w+)\]', o)[0]
+        mounts_cmd = "cat /proc/mounts | grep /dev/%s" % dev
+        s, o = session.cmd_status_output(mounts_cmd)
+        if s:
+            s, o = session.cmd_status_output('mount /dev/%s /mnt' % dev)
+            if s:
+                test.error("Fail to mount /dev/%s, output: %s" % (dev, o))
+            mp = "/mnt"
+        else:
+            mp = re.findall(r'/dev/%s\d*\s+(\S+)\s+' % dev, o)[0]
+        return mp
+
+    def _usb_stick_io(mount_point):
+        """
+        Do I/O operations on passthrough USB stick
+        """
+        error_context.context("Read and write on USB stick ", logging.info)
+        testfile = os.path.join(mount_point, 'testfile')
+        iozone_cmd = params.get("iozone_cmd",
+                                " -a -I -r 64k -s 1m -i 0 -i 1 -f %s")
+        iozone_test.run(iozone_cmd % testfile)
+
+    usbredirdev_name = params["usbredirdev_name"]
+    usbredir_params = params.object_params(usbredirdev_name)
+    backend = usbredir_params.get('chardev_backend', 'spicevmc')
+    if backend not in ('spicevmc', 'tcp_socket'):
+        test.error("Unsupported char device backend type: %s" % backend)
+
+    option = params.get("option")
+    vendorid = params["usbredir_vendorid"]
+    productid = params["usbredir_productid"]
+    timeout = params.get("wait_timeout", 600)
+    lsusb_cmd = "lsusb -v -d %s:%s" % (vendorid, productid)
+    usb_stick = "Mass Storage" in process.getoutput(lsusb_cmd)
+    rv_binary = params.get('rv_binary', 'remote-viewer')
+
+    error_context.context("Check host configurations", logging.info)
+    s, o = _host_config_check()
+    if not s:
+        test.error(o)
+
+    error_context.context("Preprocess VM", logging.info)
+    _usbredir_preprocess()
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```

Comment:
Suggest move L205~206 into `_usbredir_preprocess`, and:
```
    def prepare_spicevmc():
        usbredir_preprocess()   --> launch_vm()
        start_spice_redirection()
        vm.resume()

    if backend == 'spicevmc':
        prepare_spicevmc()
```

Reply:
The variable `vm` is used so many places in the `run` function. If we move line 205-206 into `_usbredir_preprocess` as you mentioned above, we need to return it explicitly from `_usbredir_preprocess` and then from `prepare_spicevmc`. So, it's better to leave it in the highest scope in the `run` function. Thanks!

Reply:
You are right.  Return a vm object is acceptable for me, which means both host and vm env is ready for the testing in next step,  How about:
```
    def prepare_spicevmc():
        usbredir_preprocess()   --> launch_vm()
        vm = env.get_vm(params["main_vm"])
        vm.verify_alive()
        start_spice_redirection()
        vm.resume()
        return vm

    if backend == 'spicevmc':
        prepare_spicevmc()
```

Reply:
Then we need to get `vm` object, status and error output of `_start_spice_redirection()` from prepare_spicevmc().

Reply:
It is not must to handle all the error in main function. I am ok with raise exception in `start_spice_redirection`.

Reply:
My point is to put the client and server in one function. For different chardev backend, go to different workflow.

Reply:
I updated as we talked, please help review again, thanks! @yanan-fu